cou/steps/analyze.py

```diff
@@ -0,0 +1,200 @@
+        for charm in charms:
+            openstack_charms.add(charm)
+
+    for app, app_status in status.items():
+        charm = extract_charm_name(app_status.charm)
+        if charm in openstack_charms:
+            os_versions[app] = get_current_os_versions((app, charm))
+
+    logging.debug(os_versions)
+    return os_versions
+
+
+def extract_app_channel_and_origin(app: str) -> Tuple:
+    """Extract application channel and origin (cs or ch) by the juju status."""
+    app_status = get_application_status(app)
+    return app_status.get("charm-channel"), app_status.get("charm").split(":")[0]
+
+
+def extract_os_charm_config(app: str) -> str:
+    """Extract application configuration for openstack-origin."""
+    app_config = model.get_application_config(app)
+    for origin in ("openstack-origin", "source"):
+        if app_config.get(origin):
+            return app_config.get(origin).get("value")
+
+    logging.warning("Failed to get origin for %s, no origin config found", app)
+    return ""
+
+
+def check_os_versions(os_versions: Dict[str, defaultdict[Any, Set]]) -> Tuple:
+    """Check the consistency of OpenStack version on the deployment."""
+    versions = defaultdict(set)
+    upgrade_units = defaultdict(set)
+    upgrade_charms = defaultdict(set)
+
+    for app, os_release_units in os_versions.items():
+        os_version_units = set(os_release_units.keys())
+        for os_version_unit in os_version_units:
+            versions[os_version_unit].add(app)
+        if len(os_version_units) > 1:
+            logging.warning("Units from %s are not in the same openstack version", app)
+            os_sequence = sorted(
+                list(os_version_units),
+                key=lambda release: CompareOpenStack(release),  # pylint: disable=W0108
+            )
+            for os_release in os_sequence[:-1]:
+                next_release = determine_next_openstack_release(os_release)[1]
+                upgrade_units[next_release].update(os_release_units[os_release])
+                logging.warning(
+                    "upgrade units: %s from: %s to %s",
+                    os_release_units[os_release],
+                    os_release,
+                    next_release,
+                )
+
+    if len(versions) > 1:
+        logging.warning("Charms are not in the same openstack version")
+        os_sequence = sorted(
+            versions.keys(), key=lambda release: CompareOpenStack(release)  # pylint: disable=W0108
+        )
+        for os_release in os_sequence[:-1]:
+            next_release = determine_next_openstack_release(os_release)[1]
+            upgrade_charms[next_release].update(versions[os_release])
+            logging.warning(
+                "upgrade charm: %s from: %s to %s", versions[os_release], os_release, next_release
+            )
+
+    else:
+        actual_release = list(versions)[0]
+        next_release = determine_next_openstack_release(actual_release)[1]
```

Comment:
Can we make function to return actual and next release? This function is too long and I do not think that using `[0]` and `[1]` like this is easy to understand.

Reply:
`determine_next_openstack_release` comes from zaza, so I think it's better not change it.